ignite/engine/__init__.py

```diff
@@ -19,16 +20,15 @@ def _prepare_batch(batch: Sequence[torch.Tensor], device: Optional[Union[str, to
     """Prepare batch for training: pass to a device with options.
 
     """
-    x, y = batch
-    return (convert_tensor(x, device=device, non_blocking=non_blocking),
-            convert_tensor(y, device=device, non_blocking=non_blocking))
+    return convert_tensor(batch, device=device, non_blocking=non_blocking)
 
 
 def create_supervised_trainer(model: torch.nn.Module, optimizer: torch.optim.Optimizer,
                               loss_fn: Union[Callable, torch.nn.Module],
                               device: Optional[Union[str, torch.device]] = None, non_blocking: bool = False,
                               prepare_batch: Callable = _prepare_batch,
-                              output_transform: Callable = lambda x, y, y_pred, loss: loss.item()) -> Engine:
+                              output_transform: Callable = lambda x, y, y_pred, loss: loss.item(),
+                              get_data: Callable = lambda batch: (batch[0], batch[1])) -> Engine:
```

Comment:
What is the difference between `get_data` and old `prepare_batch` ?

Reply:
In the previous implementation `prepare_batch` was responsible for pushing it to the correct device and dtype and for ensuring it's a tuple. This has now changed.

Now the `prepare_batch` supports arbitrary types (and still tuples), since it has no longer an explicit constraint on tuples, but the `get_data` function defines how we actually build the tuple we need out of an arbitrary data source.